File: askbot/utils/markdown_plugins/video_embed.py
"""
Video embedding plugin for markdown-it-py.

Syntax:
    @[youtube](video_id)
    @[vimeo](video_id)
    @[dailymotion](video_id)

Renders as iframe embed for supported services.

Based on: https://github.com/CenterForOpenScience/markdown-it-video
"""

import re
from markdown_it import MarkdownIt
from markdown_it.rules_inline import StateInline


VIDEO_SERVICES = {
    'youtube': {
        'url': 'https://www.youtube.com/embed/{0}',
        'width': 640,
        'height': 390,
    },
    'vimeo': {
        'url': 'https://player.vimeo.com/video/{0}',
        'width': 640,
        'height': 360,
    },
    'dailymotion': {
        'url': 'https://www.dailymotion.com/embed/video/{0}',
        'width': 640,
        'height': 360,
    },
}


def video_embed_rule(state: StateInline, silent: bool) -> bool:
    """
    Parse video embed syntax: @[service](video_id)

    Returns True if pattern matches and token created.
    """
    pos = state.pos
    maximum = state.posMax

    # Must start with @[
    if state.src[pos:pos+2] != '@[':
        return False

    # Find closing ]
    service_start = pos + 2
    service_end = state.src.find(']', service_start)

    if service_end == -1 or service_end >= maximum:
        return False

    service = state.src[service_start:service_end].strip().lower()

    # Check if it's a supported service
    if service not in VIDEO_SERVICES:
        return False

    # Must have opening (
    if service_end + 1 >= maximum or state.src[service_end + 1] != '(':
        return False

    # Find closing )
    id_start = service_end + 2
    id_end = state.src.find(')', id_start)

    if id_end == -1 or id_end >= maximum:
        return False

    video_id = state.src[id_start:id_end].strip()

    # Validate video ID (alphanumeric, dashes, underscores)
    if not re.match(r'^[a-zA-Z0-9_-]+$', video_id):
        return False

    if not silent:
        token = state.push('video_embed', '', 0)
        token.meta = {
            'service': service,
            'id': video_id,
        }
        token.markup = state.src[pos:id_end+1]

    state.pos = id_end + 1
    return True


def render_video_embed(self, tokens, idx, options, env):
    """
    Render video embed token as iframe.
    """
    token = tokens[idx]
    service = token.meta['service']
    video_id = token.meta['id']

    config = VIDEO_SERVICES[service]
    url = config['url'].format(video_id)
    width = config['width']
    height = config['height']

    # Security: Only allow whitelisted domains
    # HTML escaping handled by renderer

    iframe = (
        f'<div class="video-embed video-embed-{service}">'
        f'<iframe '
        f'width="{width}" '
        f'height="{height}" '
        f'src="{url}" '
        f'frameborder="0" '
        f'allowfullscreen '
        f'loading="lazy"'
        f'></iframe>'
        f'</div>'
    )

    return iframe


def video_embed_plugin(md: MarkdownIt) -> MarkdownIt:
    """
    Plugin to enable video embedding in markdown.

    Usage:
        md = MarkdownIt()
        md.use(video_embed_plugin)
    """
    # Register inline rule before 'link' to give it priority
    md.inline.ruler.before('link', 'video_embed', video_embed_rule)

    # Register renderer
    md.add_render_rule('video_embed', render_video_embed)

    return md

File: askbot/media/wmd/markdown-it/plugins/video_embed.js
/**
 * Video embedding plugin for markdown-it.
 *
 * Syntax:
 *     @[youtube](video_id)
 *     @[vimeo](video_id)
 *     @[dailymotion](video_id)
 *
 * Renders as iframe embed for supported services.
 *
 * Based on: https://github.com/CenterForOpenScience/markdown-it-video
 * Ported from: askbot/utils/markdown_plugins/video_embed.py
 */
(function(window) {
    'use strict';

    var VIDEO_SERVICES = {
        youtube: {
            url: 'https://www.youtube.com/embed/{0}',
            width: 640,
            height: 390
        },
        vimeo: {
            url: 'https://player.vimeo.com/video/{0}',
            width: 640,
            height: 360
        },
        dailymotion: {
            url: 'https://www.dailymotion.com/embed/video/{0}',
            width: 640,
            height: 360
        }
    };

    // Regex for valid video ID: alphanumeric, dashes, underscores
    var VIDEO_ID_REGEX = /^[a-zA-Z0-9_-]+$/;

    /**
     * Parse video embed syntax: @[service](video_id)
     *
     * Returns true if pattern matches and token created.
     */
    function videoEmbedRule(state, silent) {
        var pos = state.pos;
        var max = state.posMax;
        var src = state.src;

        // Must start with @[
        if (src.charCodeAt(pos) !== 0x40 /* @ */ ||
            src.charCodeAt(pos + 1) !== 0x5B /* [ */) {
            return false;
        }

        // Find closing ]
        var serviceStart = pos + 2;
        var serviceEnd = src.indexOf(']', serviceStart);

        if (serviceEnd === -1 || serviceEnd >= max) {
            return false;
        }

        var service = src.slice(serviceStart, serviceEnd).trim().toLowerCase();

        // Check if it's a supported service
        if (!VIDEO_SERVICES.hasOwnProperty(service)) {
            return false;
        }

        // Must have opening (
        if (serviceEnd + 1 >= max || src.charCodeAt(serviceEnd + 1) !== 0x28 /* ( */) {
            return false;
        }

        // Find closing )
        var idStart = serviceEnd + 2;
        var idEnd = src.indexOf(')', idStart);

        if (idEnd === -1 || idEnd >= max) {
            return false;
        }

        var videoId = src.slice(idStart, idEnd).trim();

        // Validate video ID (alphanumeric, dashes, underscores)
        if (!VIDEO_ID_REGEX.test(videoId)) {
            return false;
        }

        if (!silent) {
            var token = state.push('video_embed', '', 0);
            token.meta = {
                service: service,
                id: videoId
            };
            token.markup = src.slice(pos, idEnd + 1);
        }

        state.pos = idEnd + 1;
        return true;
    }

    /**
     * Render video embed token as iframe.
     */
    function renderVideoEmbed(tokens, idx, options, env, self) {
        var token = tokens[idx];
        var service = token.meta.service;
        var videoId = token.meta.id;

        var config = VIDEO_SERVICES[service];
        var url = config.url.replace('{0}', videoId);
        var width = config.width;
        var height = config.height;

        // Security: Only allow whitelisted domains (enforced by VIDEO_SERVICES)
        // HTML escaping handled by using safe string construction

        return '<div class="video-embed video-embed-' + service + '">' +
            '<div class="video-embed-wrapper">' +
            '<iframe ' +
            'src="' + url + '" ' +
            'frameborder="0" ' +
            'allowfullscreen ' +
            'loading="lazy"' +
            '></iframe>' +
            '</div>' +
            '</div>';
    }

    /**
     * Plugin to enable video embedding in markdown.
     *
     * Usage:
     *     var md = window.markdownit();
     *     md.use(window.markdownitVideoEmbed);
     */
    function videoEmbedPlugin(md) {
        // Register inline rule before 'link' to give it priority
        md.inline.ruler.before('link', 'video_embed', videoEmbedRule);

        // Register renderer
        md.renderer.rules.video_embed = renderVideoEmbed;

        return md;
    }

    // Export for browser and module environments
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = videoEmbedPlugin;
    } else {
        window.markdownitVideoEmbed = videoEmbedPlugin;
    }

})(typeof window !== 'undefined' ? window : this);
